File: backend/app/utils/timetable_conflicts.py
from datetime import time

from sqlalchemy.orm import Session

from app.models.timetable import Timetable
from app.utils.datetime_format import parse_time_value
from app.utils.timetable_schedule_policy import (
    validate_class_schedule
)


def times_overlap(
    start_a: time,
    end_a: time,
    start_b: time,
    end_b: time
) -> bool:
    return start_a < end_b and start_b < end_a


def validate_timetable_schedule(
    db: Session,
    day_of_week: str,
    start_time,
    end_time,
    room_number: str,
    instructor_name: str,
    course_id: int,
    exclude_timetable_id: int = None
):
    start = (
        start_time
        if isinstance(start_time, time)
        else parse_time_value(start_time)
    )
    end = (
        end_time
        if isinstance(end_time, time)
        else parse_time_value(end_time)
    )

    if end <= start:
        raise ValueError(
            "End time must be after start time"
        )

    policy_error = validate_class_schedule(
        day_of_week,
        start,
        end
    )

    if policy_error:
        raise ValueError(policy_error)

    query = db.query(Timetable).filter(
        Timetable.day_of_week == day_of_week
    )

    if exclude_timetable_id is not None:
        query = query.filter(
            Timetable.id != exclude_timetable_id
        )

    for existing in query.all():
        existing_start = parse_time_value(
            existing.start_time
        )
        existing_end = parse_time_value(
            existing.end_time
        )

        if not times_overlap(
            start,
            end,
            existing_start,
            existing_end
        ):
            continue

        if existing.room_number == room_number:
            raise ValueError(
                f"Room {room_number} is already booked "
                f"on {day_of_week} during this time"
            )

        if existing.instructor_name == instructor_name:
            raise ValueError(
                f"{instructor_name} is not available "
                f"on {day_of_week} during this time"
            )

        if existing.course_id == course_id:
            raise ValueError(
                "This course is already scheduled "
                "on this day during this time"
            )

    for existing in db.query(Timetable).filter(
        Timetable.course_id == course_id
    ).all():
        if (
            exclude_timetable_id is not None and
            existing.id == exclude_timetable_id
        ):
            continue

        if existing.instructor_name == instructor_name:
            raise ValueError(
                f"{instructor_name} already has a class section "
                "for this course"
            )

File: frontend/src/utils/timetableConflicts.js
import { validateClassSchedule } from "../constants/timetableSchedule"

function toMinutes(value) {
  const [hour, minute] = value.split(":").map(Number)
  return hour * 60 + minute
}

export function timesOverlap(startA, endA, startB, endB) {
  const startMinutesA = toMinutes(startA)
  const endMinutesA = toMinutes(endA)
  const startMinutesB = toMinutes(startB)
  const endMinutesB = toMinutes(endB)

  return (
    startMinutesA < endMinutesB &&
    startMinutesB < endMinutesA
  )
}

export function getOccupiedRooms({
  entries,
  dayOfWeek,
  startTime,
  endTime,
  excludeId = null
}) {
  if (!dayOfWeek || !startTime || !endTime) {
    return new Set()
  }

  if (toMinutes(endTime) <= toMinutes(startTime)) {
    return new Set()
  }

  const occupied = new Set()

  for (const entry of entries) {
    if (excludeId && entry.id === excludeId) {
      continue
    }

    if (entry.day_of_week !== dayOfWeek) {
      continue
    }

    if (
      !timesOverlap(
        entry.start_time,
        entry.end_time,
        startTime,
        endTime
      )
    ) {
      continue
    }

    occupied.add(entry.room_number)
  }

  return occupied
}

export function getAvailableRooms({
  allRooms,
  entries,
  dayOfWeek,
  startTime,
  endTime,
  excludeId = null
}) {
  const occupied = getOccupiedRooms({
    entries,
    dayOfWeek,
    startTime,
    endTime,
    excludeId
  })

  return allRooms.filter((room) => !occupied.has(room))
}

export function findInstructorConflict({
  entries,
  dayOfWeek,
  startTime,
  endTime,
  instructorName,
  excludeId = null
}) {
  if (
    !dayOfWeek ||
    !startTime ||
    !endTime ||
    !instructorName
  ) {
    return null
  }

  if (toMinutes(endTime) <= toMinutes(startTime)) {
    return null
  }

  for (const entry of entries) {
    if (excludeId && entry.id === excludeId) {
      continue
    }

    if (entry.day_of_week !== dayOfWeek) {
      continue
    }

    if (
      !timesOverlap(
        entry.start_time,
        entry.end_time,
        startTime,
        endTime
      )
    ) {
      continue
    }

    if (entry.instructor_name === instructorName) {
      return (
        `${instructorName} is not available ` +
        `on ${dayOfWeek} during this time`
      )
    }
  }

  return null
}

export function findTimetableConflict({
  entries,
  dayOfWeek,
  startTime,
  endTime,
  roomNumber,
  instructorName,
  courseId,
  excludeId = null
}) {
  if (!dayOfWeek || !startTime || !endTime) {
    return null
  }

  if (toMinutes(endTime) <= toMinutes(startTime)) {
    return "End time must be after start time"
  }

  const policyError = validateClassSchedule(
    dayOfWeek,
    startTime,
    endTime
  )

  if (policyError) {
    return policyError
  }

  for (const entry of entries) {
    if (excludeId && entry.id === excludeId) {
      continue
    }

    if (entry.day_of_week !== dayOfWeek) {
      continue
    }

    if (
      !timesOverlap(
        entry.start_time,
        entry.end_time,
        startTime,
        endTime
      )
    ) {
      continue
    }

    if (entry.room_number === roomNumber) {
      return (
        `Room ${roomNumber} is already booked ` +
        `on ${dayOfWeek} during this time`
      )
    }

    if (entry.instructor_name === instructorName) {
      return (
        `${instructorName} is not available ` +
        `on ${dayOfWeek} during this time`
      )
    }

    if (entry.course_id === Number(courseId)) {
      return (
        "This course is already scheduled " +
        "on this day during this time"
      )
    }
  }

  if (Number(courseId) && instructorName) {
    for (const entry of entries) {
      if (excludeId && entry.id === excludeId) {
        continue
      }

      if (
        entry.course_id === Number(courseId) &&
        entry.instructor_name === instructorName
      ) {
        return (
          `${instructorName} already has a class section ` +
          "for this course"
        )
      }
    }
  }

  return null
}
